source/utilities.cc

```diff
@@ -2535,7 +2535,67 @@ namespace aspect
       return operator_list;
     }
 
+    template <int dim>
+    GeneralizedPoint<dim>::GeneralizedPoint(Point<dim> position, const GeometryModel::Interface<dim> *geometry_model)
+    {
+      cartesian_position = position;
+      coordinate_system = geometry_model->natural_coordinate_system();
+      coordinates = geometry_model->cartesian_to_natural_coordinates(cartesian_position);
+    }
+
+    template <int dim>
+    Point<dim> &GeneralizedPoint<dim>::get_position()
+    {
+      return cartesian_position;
+    }
+
+    template <int dim>
+    std_cxx11::array<double,dim> &GeneralizedPoint<dim>::get_coordinates()
+    {
+      return coordinates;
+    }
+
+    template <int dim>
+    std_cxx11::array<double,dim-1> GeneralizedPoint<dim>::get_surface_coordinates() const
+    {
+      if (coordinate_system == Coordinates::CoordinateSystem::cartesian)
+        {
+          if (dim == 2)
+            {
+              std_cxx11::array<double,1> a {coordinates[0]};
```

Comment:
This only works in C++11, which we aren't allowed to use yet :-( I'm afraid you'll have to initialize components individually.